### DISCUSSION TITLE
.git-blame-ignore-revs

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#13626] [General]
When discussing code refactoring over the past few years (for various reasons: code formatting, moving files, etc.) it inevitably leads to concerns about making .git blame unusable. That holds back certain things like reformatting the code base.

It seems a new feature (last few years) is available in not only git itself but also GitHub that removes this concern. I'm sharing it here so all are aware for the inevitable next discussion :)

 * https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/ : While the article says .git-blame-ignore-revs is beta it has been released to everyone according to the issue discussion.

@grokys @kekekeks

**Comment by maxkatz6** [comment]
That's an interesting one. But I assume it's supposed to be used with manual changes and refactorings. And not with automated formatting. Otherwise, this git-blame-ignore-revs file would be huge.

**Reply by maxkatz6** [reply]
We still should start using it for refactorings and reformattings, IMO.

**Reply by robloo** [reply]
> manual changes and refactorings. And not with automated formatting.

It was actually first implemented for big automated code refactoring that mess literally everything up otherwise. I read somewhere a while ago (of course I can't find it now) where it was used on some large project when they applied automated code formatting to the entire code base.

> We still should start using it for refactorings and reformattings, IMO.

The .git-blame-ignore-revs file could grow pretty large if it was used for all refactoring commits indiscriminately. IMO big formatting/refactoring should be grouped together in one commit to reduce the number. The big Avalonia.Base refactoring could have been squashed and put in this file for example.

It would be great to reorganize the controls directory someday in one squashed commit -- again perfect for this file.

So just to keep the file size smaller (avoid hundreds of line entries) it's probably best to still plan refactoring together.

**Reply by robloo** [reply]
I remembered the example I read about previously was Django. That project ran a code formatter (black?) over the code base then used .git-blame-ignore-revs to avoid messing up the blame history. There was a blog post and associated GitHub discussion I believe. It was all done it a huge commit.
